api/controllers/UserController.js

```diff
@@ -234,6 +235,318 @@ module.exports = {
+                }
+                  completed_tasks++;
+                  if (completed_tasks >= number_of_tasks) {
+                    return res.json('ok', 200);
+                  }
+              });
+          }
+        }
+      );
+    }
+
+    //Update the AutoModerator config with the user's friend codes
+    var updateAutomod = function (subreddit, friend_codes) {
+      Reddit.getWikiPage(
+        req.user.redToken,
+        subreddit,
+        'config/automoderator',
+        function (err, current_config) {
+          if (err) {
+            console.log(err);
+            return res.json('Error retrieving /r/' + subreddit + ' AutoModerator config', 500);
+          }
+          else {
+            var lines = current_config.data.content_md.split("\r\n");
+            var fclist_indices = [lines.indexOf("#FCList1") + 1, lines.indexOf("#FCList2") + 1];
+            if (fclist_indices.indexOf(0) != -1) {
+              console.log("Error: Could not find #FCList tags in /r/" + subreddit + " AutoModerator config");
+              return res.json('Error parsing /r/' + subreddit + ' AutoModerator config', 500);
+            }
+            try {
+              for (var listno = 0; listno < fclist_indices.length; listno++) {
+                var before_bracket = lines[fclist_indices[listno]].substring(0,lines[fclist_indices[listno]].indexOf("]"));
+                for (var i = 0; i < friend_codes.length; i++) {
+                  before_bracket += ", \"" + friend_codes[i] + "\"";
+                }
+                lines[fclist_indices[listno]] = before_bracket + "]";
+              }
+            }
+            catch (err) {
+              console.log('Error parsing /r/" + subreddit + " AutoModerator config');
+              return res.json('Error parsing /r/" + subreddit + " AutoModerator config', 500);
+            }
+            var content = lines.join("\r\n");
+            Reddit.editWikiPage(
+              req.user.redToken,
+              subreddit,
+              'config/automoderator',
+              content,
+              'FlairHQ: Updated banned friend codes',
+              function (err, response) {
+                if (err) {
+                  console.log(err);
+                  return res.json('Failed to update /r/' + subreddit + ' AutoModerator config', 500);
+                } else {
+                    console.log("Added /u/" + req.params.username + "'s friend codes to /r/" + subreddit + " AutoModerator blacklist");
+                    Application.destroy({id: req.allParams().id}).exec(function (err, app) {
+                      if (err) {
+                        console.log(err);
+                        return res.json('Error while updating /r/' + subreddit + 'AutoModerator config', 500);
+                      }
+                        completed_tasks++;
+                        if (completed_tasks >= number_of_tasks) {
+                          return res.json('ok', 200);
```

Comment:
Aha, I see what you are doing here. Good job, but can you use promises instead?

So, make all of these functions return a promise, and do Q.all on them, so that we know when they are all done, and it's much nicer.

If you don't know about promises, give me a shout and I can go over them quickly with you. They are very odd and confusing to start with, but they are beautiful after you know how to work them.